// @workspace/db/src/schema/studies.ts
import { pgTable, text, serial, integer, timestamp } from "drizzle-orm/pg-core";
import { createInsertSchema } from "drizzle-zod";
import { z } from "zod/v4";

export const studiesTable = pgTable("studies", {
  id: serial("id").primaryKey(),
  shareToken: text("share_token").notNull().unique(),
  patientName: text("patient_name").notNull(),
  patientId: text("patient_id").notNull(),
  patientDob: text("patient_dob"),
  studyDescription: text("study_description"),
  studyDate: text("study_date"),
  createdAt: timestamp("created_at").defaultNow().notNull(),
});

export const insertStudySchema = createInsertSchema(studiesTable).omit({ id: true, createdAt: true });
export type InsertStudy = z.infer<typeof insertStudySchema>;
export type Study = typeof studiesTable.$inferSelect;

export const imagesTable = pgTable("images", {
  id: serial("id").primaryKey(),
  studyId: integer("study_id").notNull().references(() => studiesTable.id, { onDelete: "cascade" }),
  filename: text("filename").notNull(),
  storagePath: text("storage_path").notNull(),
  imageOrder: integer("image_order").notNull().default(0),
  createdAt: timestamp("created_at").defaultNow().notNull(),
});

export const insertImageSchema = createInsertSchema(imagesTable).omit({ id: true, createdAt: true });
export type InsertImage = z.infer<typeof insertImageSchema>;
export type Image = typeof imagesTable.$inferSelect;

export const notesTable = pgTable("notes", {
  id: serial("id").primaryKey(),
  studyId: integer("study_id").notNull().references(() => studiesTable.id, { onDelete: "cascade" }),
  content: text("content").notNull(),
  author: text("author"),
  createdAt: timestamp("created_at").defaultNow().notNull(),
});

export const insertNoteSchema = createInsertSchema(notesTable).omit({ id: true, createdAt: true });
export type InsertNote = z.infer<typeof insertNoteSchema>;
export type Note = typeof notesTable.$inferSelect;
